src/traces/parcats/calc.js

```diff
@@ -0,0 +1,526 @@
+var colorscaleCalc = require('../../components/colorscale/calc');
+var parcatConstants = require('./constants');
+var filterUnique = require('../../lib/filter_unique.js');
+var Drawing = require('../../components/drawing');
+var Lib = require('../../lib');
+
+
+function visible(dimension) { return !('visible' in dimension) || dimension.visible; }
+
+// Exports
+// =======
+/**
+ * Create a wrapped ParcatsModel object from trace
+ *
+ * Note: trace defaults have already been applied
+ * @param {Object} gd
+ * @param {Object} trace
+ * @return {Array.<ParcatsModel>}
+ */
+module.exports = function calc(gd, trace) {
+
+    // Process inputs
+    // --------------
+    if(trace.dimensions.filter(visible).length === 0) {
+        // No visible dimensions specified. Nothing to compute
+        return [];
+    }
+
+    // Compute unique information
+    // --------------------------
+    // UniqueInfo per dimension
+    var uniqueInfoDims = trace.dimensions.filter(visible).map(function(dim) {
+        var categoryValues;
+        if(dim.categoryorder === 'trace') {
+            // Use order of first occurrence in trace
+            categoryValues = null;
+        } else if(dim.categoryorder === 'array') {
+            // Use categories specified in `categoryarray` first, then add extra to the end in trace order
+            categoryValues = dim.categoryarray;
+        } else {
+            // Get all categories up front so we can order them
+            // Should we check for numbers as sort numerically?
+            categoryValues = filterUnique(dim.values).sort();
+            if(dim.categoryorder === 'category descending') {
+                categoryValues = categoryValues.reverse();
+            }
+        }
+        return getUniqueInfo(dim.values, categoryValues);
+    });
+
+    // Process counts
+    // --------------
+    var counts,
+        count,
+        totalCount;
+    if(Lib.isArrayOrTypedArray(trace.counts)) {
+        counts = trace.counts;
+    } else {
+        counts = [trace.counts];
+    }
+
+    // Validate dimension display order
+    // --------------------------------
+    validateDimensionDisplayInds(trace);
+
+    // Validate category display order
+    // -------------------------------
+    trace.dimensions.filter(visible).forEach(function(dim, dimInd) {
+        validateCategoryProperties(dim, uniqueInfoDims[dimInd]);
+    });
+
+    // Handle path colors
+    // ------------------
+    var line = trace.line;
+    var markerColorscale;
+
+    // Process colorscale
+    if(line) {
+        if(hasColorscale(trace, 'line')) {
+            colorscaleCalc(trace, trace.line.color, 'line', 'c');
+        }
+        markerColorscale = Drawing.tryColorscale(line);
+    } else {
+        markerColorscale = Lib.identity;
+    }
+
+    // Build color generation function
+    function getMarkerColorInfo(index) {
+        var value;
+        if(!line) {
+            value = parcatConstants.defaultColor;
```

Comment:
isn't this case ruled out (and the default supplied) by `supplyDefaults`? And it looks like this line is the only use of anything in `parcatConstants`, so perhaps we can 🔪 that?

Reply:
Yes, thanks for catching these holdovers from before I understood how `supplyDefaults` worked 🙂 Done in eba1b4a